common/tablefunctions.lua

```diff
@@ -104,11 +156,170 @@ if not table.invert then
 	end
 end
 
-
 if not table.append then
 	function table.append(appendTarget, appendData)
 		for _, value in pairs(appendData) do
 			table.insert(appendTarget, value)
 		end
 	end
 end
+
+if not table.count then
+	---Count the number of values in table.
+	---Note that this always works, whereas the default length operator (#table)
+	---only works if the table is a Lua sequence (i.e. indexes form a contiguous
+	---sequence starting from 1).
+	---@param tbl table
+	---@return integer
+	function table.count(tbl)
+		local count = 0
+		for _ in pairs(tbl) do
+			count = count + 1
+		end
+		return count
+	end
+end
+
+if not table.findKey then
+	---Find key of value in table.
+	---Will always return the first key found, no matter if the table contains
+	---multiple instances of the value.
+	---@param tbl table
+	---@param value any
+	---@return any # key if found, nil otherwise
+	function table.findKey(tbl, value)
+		for key, v in pairs(tbl) do
+			if v == value then
+				return key
+			end
+		end
+		return nil
+	end
+end
+
+if not table.contains then
+	---Check if value is in table.
+	---@param tbl table
+	---@param value boolean
+	function table.contains(tbl, value)
+		return table.findKey(tbl, value) ~= nil
+	end
+end
+
+if not table.removeValue then
```

Comment:
consider also `removeValueIf` with a conditional, or `removeAllValue` which is just `while table.removeValue(x,y) do end`

Reply:
How would `removeValueIf` be more useful than just doing
```
if condition then
  table.removeValue(someTable, value)
end
```

 Wouldn't `removeAllValue` just be `someTable = {}`?

Reply:
I don't see value either in `removeValueIf` but I do see value in having `removeValue` being able to handle removing all instances of a value in table. I could actually have that be the default behaviour if we wish to.

Reply:
> How would `removeValueIf` be more useful than just doing `if condition`

Because that is not what this would do.
```lua
local t = {a = 1, b = 2, c = 3, d = 4, e = 5}
t:removeIf(function(x) return x % 2 == 1 end) -- remove if odd
-- t is now {b = 2, d = 4}
```

> Wouldn't `removeAllValue` just be `someTable = {}`?

No, it removes all of the given value (unlike the existing one which removes one arbitrary instance of it):
```lua
local t = { a = 1, b = 2, c = 1, d = 2 }
t:removeValue(2) -- removes one arbitrary entry with value 2: maybe b, or maybe d
t:removeAllValue(2) -- removes all entries with value 2: both b and d
```

> an optional flag passed to the function

I dislike magic bool flags. Consider:
* `shallowCopy(x)` and `deepCopy(x)`. Which one is the shallow copy and which is the deep copy?
* `copy(x, true)` and `copy(x, false)`. Which one is the shallow copy and which is the deep copy?

Reply:
Added both of them, PTAL :)